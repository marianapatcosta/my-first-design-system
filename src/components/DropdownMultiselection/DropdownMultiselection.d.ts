import React, { CSSProperties, MouseEventHandler } from 'react';
export interface DropdownMultiselectionProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
     * Dropdown's displaying text when no option is selected or if multiple options can be selected
     */
    placeholder?: string;
    /**
     * Is Dropdown disabled?
     */
    disabled?: boolean;
    /**
     * Is Dropdown songle selection?
     */
    isSingleSelection?: boolean;
    /**
     * the options that are selected
     */
    selectedOptions?: Array<any>;
    /**
     * the selected option, if dropdown only allows to select on item each time
     */
    selectedOption?: any;
    /**
     * name of the field to be displayed in the list of dropdown's options, when options prop is an array of objects
     */
    optionKey?: string;
    /**
     * options to display in the dropdown content
     */
    options: Array<any>;
    /**
     * Change handler
     */
    onOptionClick: MouseEventHandler<HTMLLIElement>;
}
export declare const DropdownMultiselection: React.FC<DropdownMultiselectionProps>;
