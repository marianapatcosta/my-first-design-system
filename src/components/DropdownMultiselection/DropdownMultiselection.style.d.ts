export declare const StyledDropdownMultiselection: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledDropdownMultiselectionHeader: import("styled-components").StyledComponent<"div", any, {
    disabled?: boolean | undefined;
}, never>;
export declare const StyledDropdownMultiselectionArrow: import("styled-components").StyledComponent<"img", any, {
    isExpanded?: boolean | undefined;
}, never>;
export declare const StyledDropdownMultiselectionOptions: import("styled-components").StyledComponent<"ul", any, {}, never>;
export declare const StyledDropdownMultiselectionOption: import("styled-components").StyledComponent<"li", any, {}, never>;
