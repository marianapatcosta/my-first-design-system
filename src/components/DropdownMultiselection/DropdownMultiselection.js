var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useRef, useState, } from 'react';
import { Chevron } from '../../assets/icons';
import { KEYBOARD_KEY } from '../../constants';
import { Checkbox } from '../Checkbox/Checkbox';
import { StyledDropdownMultiselection, StyledDropdownMultiselectionHeader, StyledDropdownMultiselectionArrow, StyledDropdownMultiselectionOptions, StyledDropdownMultiselectionOption, } from './DropdownMultiselection.style';
export var DropdownMultiselection = function (_a) {
    var placeholder = _a.placeholder, options = _a.options, optionKey = _a.optionKey, disabled = _a.disabled, isSingleSelection = _a.isSingleSelection, selectedOptions = _a.selectedOptions, selectedOption = _a.selectedOption, onOptionClick = _a.onOptionClick, props = __rest(_a, ["placeholder", "options", "optionKey", "disabled", "isSingleSelection", "selectedOptions", "selectedOption", "onOptionClick"]);
    var _b = useState(false), isExpanded = _b[0], setIsExpanded = _b[1];
    var dropdownRef = useRef(null);
    useEffect(function () {
        var handleClickOutside = function (event) {
            var element = event.target;
            if (dropdownRef.current && !(dropdownRef === null || dropdownRef === void 0 ? void 0 : dropdownRef.current.contains(element))) {
                event.preventDefault();
                setIsExpanded(false);
            }
        };
        document.addEventListener('mouseup', handleClickOutside);
        return function () { return document.removeEventListener('mouseup', handleClickOutside); };
    }, [dropdownRef]);
    var isOptionSelected = function (option) {
        if (!!(selectedOptions === null || selectedOptions === void 0 ? void 0 : selectedOptions.length)) {
            var selectedOptionsLabels = optionKey
                ? selectedOptions.map(function (option) { return option[optionKey]; })
                : selectedOptions;
            return selectedOptionsLabels === null || selectedOptionsLabels === void 0 ? void 0 : selectedOptionsLabels.includes(optionKey ? option[optionKey] : option);
        }
        return optionKey
            ? option[optionKey] === selectedOption[optionKey]
            : selectedOption === option;
    };
    var toggleDropdownExpansion = function () {
        !disabled && setIsExpanded(function (prevIsExpanded) { return !prevIsExpanded; });
    };
    var handleKeyEvent = function (event) {
        event.stopPropagation();
        if (disabled)
            return;
        if (event.key === KEYBOARD_KEY.ESCAPE_KEY) {
            return setIsExpanded(false);
        }
        if (event.key === KEYBOARD_KEY.ENTER_KEY) {
            return setIsExpanded(function (prevIsExpanded) { return !prevIsExpanded; });
        }
        if (event.key === KEYBOARD_KEY.SPACE_KEY) {
            return setIsExpanded(true);
        }
    };
    var getPlaceholder = function () {
        if (!isSingleSelection)
            return placeholder;
        return optionKey ? selectedOption[optionKey] : selectedOption;
    };
    return (_jsxs(StyledDropdownMultiselection, __assign({}, props, { ref: dropdownRef }, { children: [_jsxs(StyledDropdownMultiselectionHeader, __assign({ disabled: disabled, role: 'button', tabIndex: disabled ? -1 : 0, "aria-label": "Select " + placeholder, "aria-expanded": isExpanded, onClick: toggleDropdownExpansion, onKeyDown: handleKeyEvent }, { children: [_jsx("div", { children: getPlaceholder() || placeholder }, void 0),
                    _jsx(StyledDropdownMultiselectionArrow, { isExpanded: isExpanded, alt: 'chevron-icon', src: Chevron }, void 0)] }), void 0),
            isExpanded && (_jsx(StyledDropdownMultiselectionOptions, __assign({ "data-testid": 'dropdown-options' }, { children: options.map(function (option, index) { return (_jsx(StyledDropdownMultiselectionOption, { children: _jsx(Checkbox, { disabled: disabled, onChange: function () { return onOptionClick(option); }, checked: isOptionSelected(option), label: optionKey ? option[optionKey] : option }, void 0) }, index * Math.random())); }) }), void 0))] }), void 0));
};
DropdownMultiselection.defaultProps = {
    id: '',
    className: '',
    style: {},
    placeholder: '',
    disabled: false,
    isSingleSelection: false,
    options: [],
    selectedOptions: [],
    selectedOption: '',
    optionKey: '',
};
