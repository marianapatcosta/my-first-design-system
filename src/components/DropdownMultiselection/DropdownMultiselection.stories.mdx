import { Meta } from '@storybook/addon-docs/blocks';
import { DropdownMultiselection } from './DropdownMultiselection';

<Meta
  title='Design System|DropdownMultiselection'
  component={DropdownMultiselection}
/>

# DropdownMultiselection

Simple Dropdown component that allows to select multiple options

## Usage

Use dropdown DropdownMultiselection when there are more than two options available and the user should pick at least one of them.
The options can be primitives or objects. `optionKey` prop indicates which field should be displayed, in dropdown options and selected
option when options are of type Object. If only one option should be selected, the props `isSingleSelection`and `selectedOption`
should be provided; otherwise, if multiple options are allowed, `selectedOptions` prop (an array of selected options) should be provided.

- Default DropdownMultiselection:

  <Story name='Default'>
    <DropdownMultiselection
      label='select an option'
      optionKey='name'
      options={[
        { name: 'option 1' },
        { name: 'option 2' },
        { name: 'option 3' },
      ]}
      selectedOptions={[]}
    />
  </Story>

  - SingleSelection DropdownMultiselection:
    <Story name='SingleSelection'>
      <DropdownMultiselection
        label='select an option'
        optionKey='name'
        options={[
          { name: 'option 1' },
          { name: 'option 2' },
          { name: 'option 3' },
        ]}
        isSingleSelection
        selectedOption={''}
      />
    </Story>

## Other important notes

not applicable.
