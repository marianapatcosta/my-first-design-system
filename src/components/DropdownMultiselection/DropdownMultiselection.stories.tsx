import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import {
  DropdownMultiselection,
  DropdownMultiselectionProps,
} from './DropdownMultiselection';

export default {
  title: 'DropdownMultiselection',
  component: DropdownMultiselection,
  args: {
    options: [{ name: 'option 1' }, { name: 'option 2' }, { name: 'option 3' }],
    optionKey: 'name',

    label: 'select an option',
  },
} as Meta;

const Template: Story<DropdownMultiselectionProps> = (args) => (
  <DropdownMultiselection {...args} />
);

export const Default = Template.bind({});
Default.args = { selectedOptions: [] };

export const SingleSelection = Template.bind({});
SingleSelection.args = { isSingleSelection: true, selectedOption: '' };
