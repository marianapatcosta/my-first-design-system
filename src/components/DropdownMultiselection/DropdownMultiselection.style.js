var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled, { keyframes } from 'styled-components';
import { baseStyles, disabledStyles, colors } from '../../styles';
var rotate = keyframes(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  0% {\n    transform: rotateX(-90deg);\n  }\n  70% {\n    transform: rotateX(20deg);\n  }\n  100% {\n    transform: rotateX(0deg);\n  }\n"], ["\n  0% {\n    transform: rotateX(-90deg);\n  }\n  70% {\n    transform: rotateX(20deg);\n  }\n  100% {\n    transform: rotateX(0deg);\n  }\n"])));
export var StyledDropdownMultiselection = styled.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  ", "\n  display: flex;\n  flex-direction: column;\n  width: 100%;\n  position: relative;\n  max-width: 10rem;\n"], ["\n  ", "\n  display: flex;\n  flex-direction: column;\n  width: 100%;\n  position: relative;\n  max-width: 10rem;\n"])), baseStyles);
export var StyledDropdownMultiselectionHeader = styled.div(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n  padding: 0.3rem 0.625rem;\n  font-size: 95%;\n  cursor: pointer;\n  box-shadow: 0 0.063rem 0.25rem ", ";\n\n  ", "\n"], ["\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n  padding: 0.3rem 0.625rem;\n  font-size: 95%;\n  cursor: pointer;\n  box-shadow: 0 0.063rem 0.25rem ", ";\n\n  ", "\n"])), colors.shadow, function (_a) {
    var disabled = _a.disabled;
    return !!disabled && "" + disabledStyles;
});
export var StyledDropdownMultiselectionArrow = styled.img(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  width: 0.8rem;\n  ", ";\n  transition: transform 0.5s ease-in-out;\n\n  ", "\n"], ["\n  width: 0.8rem;\n  ", ";\n  transition: transform 0.5s ease-in-out;\n\n  ",
    "\n"])), colors.icon, function (_a) {
    var isExpanded = _a.isExpanded;
    return !!isExpanded &&
        "transform: rotate(180deg);\n  transition: transform 0.5s linear;";
});
export var StyledDropdownMultiselectionOptions = styled.ul(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  background-color: ", ";\n  box-shadow: 0 0.063rem 0.25rem ", ";\n  border-top: none;\n  box-sizing: border-box;\n  margin: 0;\n  padding: 0.3rem;\n  width: 100%;\n  list-style: none;\n  max-height: 12.5rem;\n  overflow: auto;\n  animation: ", " 700ms ease-in forwards;\n  -webkit-animation: ", " 700ms ease-in forwards;\n  transform-origin: top center;\n  position: absolute;\n  z-index: 1;\n  top: 2rem;\n"], ["\n  background-color: ", ";\n  box-shadow: 0 0.063rem 0.25rem ", ";\n  border-top: none;\n  box-sizing: border-box;\n  margin: 0;\n  padding: 0.3rem;\n  width: 100%;\n  list-style: none;\n  max-height: 12.5rem;\n  overflow: auto;\n  animation: ", " 700ms ease-in forwards;\n  -webkit-animation: ", " 700ms ease-in forwards;\n  transform-origin: top center;\n  position: absolute;\n  z-index: 1;\n  top: 2rem;\n"])), colors.primary, colors.shadow, rotate, rotate);
export var StyledDropdownMultiselectionOption = styled.li(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  padding: 0.3rem;\n  cursor: pointer;\n"], ["\n  padding: 0.3rem;\n  cursor: pointer;\n"])));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
