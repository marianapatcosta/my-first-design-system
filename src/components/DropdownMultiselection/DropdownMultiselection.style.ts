import styled, { keyframes } from 'styled-components';
import { baseStyles, disabledStyles, colors } from '../../styles';

const rotate = keyframes`
  0% {
    transform: rotateX(-90deg);
  }
  70% {
    transform: rotateX(20deg);
  }
  100% {
    transform: rotateX(0deg);
  }
`;

export const StyledDropdownMultiselection = styled.div`
  ${baseStyles}
  display: flex;
  flex-direction: column;
  width: 100%;
  position: relative;
  max-width: 10rem;
`;

export const StyledDropdownMultiselectionHeader = styled.div<{
  disabled?: boolean;
}>`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.3rem 0.625rem;
  font-size: 95%;
  cursor: pointer;
  box-shadow: 0 0.063rem 0.25rem ${colors.shadow};

  ${({ disabled }) => !!disabled && `${disabledStyles}`}
`;

export const StyledDropdownMultiselectionArrow = styled.img<{
  isExpanded?: boolean;
}>`
  width: 0.8rem;
  ${colors.icon};
  transition: transform 0.5s ease-in-out;

  ${({ isExpanded }) =>
    !!isExpanded &&
    `transform: rotate(180deg);
  transition: transform 0.5s linear;`}
`;

export const StyledDropdownMultiselectionOptions = styled.ul`
  background-color: ${colors.primary};
  box-shadow: 0 0.063rem 0.25rem ${colors.shadow};
  border-top: none;
  box-sizing: border-box;
  margin: 0;
  padding: 0.3rem;
  width: 100%;
  list-style: none;
  max-height: 12.5rem;
  overflow: auto;
  animation: ${rotate} 700ms ease-in forwards;
  -webkit-animation: ${rotate} 700ms ease-in forwards;
  transform-origin: top center;
  position: absolute;
  z-index: 1;
  top: 2rem;
`;

export const StyledDropdownMultiselectionOption = styled.li`
  padding: 0.3rem;
  cursor: pointer;
`;
