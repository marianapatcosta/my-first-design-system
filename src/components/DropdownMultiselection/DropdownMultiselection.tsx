import React, {
  CSSProperties,
  KeyboardEvent,
  MouseEventHandler,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Chevron } from '../../assets/icons';
import { KEYBOARD_KEY } from '../../constants';
import { Checkbox } from '../Checkbox/Checkbox';
import {
  StyledDropdownMultiselection,
  StyledDropdownMultiselectionHeader,
  StyledDropdownMultiselectionArrow,
  StyledDropdownMultiselectionOptions,
  StyledDropdownMultiselectionOption,
} from './DropdownMultiselection.style';

export interface DropdownMultiselectionProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
   * Dropdown's displaying text when no option is selected or if multiple options can be selected
   */
  placeholder?: string;
  /**
   * Is Dropdown disabled?
   */
  disabled?: boolean;
  /**
   * Is Dropdown songle selection?
   */
  isSingleSelection?: boolean;
  /**
   * the options that are selected
   */
  selectedOptions?: Array<any>;
  /**
   * the selected option, if dropdown only allows to select on item each time
   */
  selectedOption?: any;
  /**
   * name of the field to be displayed in the list of dropdown's options, when options prop is an array of objects
   */
  optionKey?: string;
  /**
   * options to display in the dropdown content
   */
  options: Array<any>;
  /**
   * Change handler
   */
  onOptionClick: MouseEventHandler<HTMLLIElement>;
}

export const DropdownMultiselection: React.FC<DropdownMultiselectionProps> = ({
  placeholder,
  options,
  optionKey,
  disabled,
  isSingleSelection,
  selectedOptions,
  selectedOption,
  onOptionClick,
  ...props
}) => {
  const [isExpanded, setIsExpanded] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handleClickOutside = (event: any) => {
      const element = event.target as Node | null;

      if (dropdownRef.current && !dropdownRef?.current.contains(element)) {
        event.preventDefault();

        setIsExpanded(false);
      }
    };
    document.addEventListener('mouseup', handleClickOutside);

    return () => document.removeEventListener('mouseup', handleClickOutside);
  }, [dropdownRef]);

  const isOptionSelected = (option: any): boolean => {
    if (!!selectedOptions?.length) {
      const selectedOptionsLabels = optionKey
        ? selectedOptions.map((option) => option[optionKey])
        : selectedOptions;
      return selectedOptionsLabels?.includes(
        optionKey ? option[optionKey] : option
      );
    }

    return optionKey
      ? option[optionKey] === selectedOption[optionKey]
      : selectedOption === option;
  };

  const toggleDropdownExpansion = () => {
    !disabled && setIsExpanded((prevIsExpanded) => !prevIsExpanded);
  };

  const handleKeyEvent = (event: KeyboardEvent<HTMLDivElement>) => {
    event.stopPropagation();
    if (disabled) return;
    if (event.key === KEYBOARD_KEY.ESCAPE_KEY) {
      return setIsExpanded(false);
    }
    if (event.key === KEYBOARD_KEY.ENTER_KEY) {
      return setIsExpanded((prevIsExpanded) => !prevIsExpanded);
    }
    if (event.key === KEYBOARD_KEY.SPACE_KEY) {
      return setIsExpanded(true);
    }
  };

  const getPlaceholder = () => {
    if (!isSingleSelection) return placeholder;
    return optionKey ? selectedOption[optionKey] : selectedOption;
  };

  return (
    <StyledDropdownMultiselection {...props} ref={dropdownRef}>
      <StyledDropdownMultiselectionHeader
        disabled={disabled}
        role='button'
        tabIndex={disabled ? -1 : 0}
        aria-label={`Select ${placeholder}`}
        aria-expanded={isExpanded}
        onClick={toggleDropdownExpansion}
        onKeyDown={handleKeyEvent}
      >
        <div>{getPlaceholder() || placeholder}</div>
        <StyledDropdownMultiselectionArrow
          isExpanded={isExpanded}
          alt='chevron-icon'
          src={Chevron}
        />
      </StyledDropdownMultiselectionHeader>
      {isExpanded && (
        <StyledDropdownMultiselectionOptions data-testid='dropdown-options'>
          {options.map((option, index) => (
            <StyledDropdownMultiselectionOption key={index * Math.random()}>
              <Checkbox
                disabled={disabled}
                onChange={() => onOptionClick(option)}
                checked={isOptionSelected(option)}
                label={optionKey ? option[optionKey] : option}
              />
            </StyledDropdownMultiselectionOption>
          ))}
        </StyledDropdownMultiselectionOptions>
      )}
    </StyledDropdownMultiselection>
  );
};

DropdownMultiselection.defaultProps = {
  id: '',
  className: '',
  style: {},
  placeholder: '',
  disabled: false,
  isSingleSelection: false,
  options: [],
  selectedOptions: [],
  selectedOption: '',
  optionKey: '',
};
