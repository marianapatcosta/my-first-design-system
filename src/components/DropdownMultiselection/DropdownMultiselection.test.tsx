import React from 'react';
import { fireEvent, getByText, render } from '@testing-library/react';
import { DropdownMultiselection } from './DropdownMultiselection';

describe('DropdownMultiselection', () => {
  const defaultProps = {
    options: [{ name: 'option 1' }, { name: 'option 2' }, { name: 'option 3' }],
    selectedOptions: [],
    optionKey: 'name',
    label: 'select an option',
    onOptionClick: () => null,
  };

  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<DropdownMultiselection {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should display label if label prop is passed', () => {
    const { container } = render(
      <DropdownMultiselection {...defaultProps} label='Dropdown' />
    );
    getByText(container, 'Dropdown');
  });

  test('should display the field of `optionKey` as option element label, if objects are passed in `options array', () => {
    const { getByText, getByLabelText } = render(
      <DropdownMultiselection
        {...defaultProps}
        label='Dropdown'
        options={[
          { label: 'option 1' },
          { label: 'option 2' },
          { label: 'option 3' },
        ]}
        optionKey='label'
      />
    );
    const dropdown = getByLabelText(/select/i) as HTMLDivElement;
    fireEvent.click(dropdown);
    getByText('option 1');
  });

  test('should handle click events and simulate selection', async () => {
    const onOptionClickMock = jest.fn();
    const { getByText, getByLabelText } = render(
      <DropdownMultiselection
        {...defaultProps}
        onOptionClick={onOptionClickMock}
      />
    );
    const dropdown = getByLabelText(/select/i) as HTMLDivElement;
    fireEvent.click(dropdown);
    const option1 = getByText(/option 1/i) as HTMLLIElement;
    fireEvent.click(option1);
    expect(onOptionClickMock).toBeCalled();
  });

  test('should set isExpanded = true onclick', () => {
    const { getByLabelText, getByTestId } = render(
      <DropdownMultiselection {...defaultProps} />
    );
    const dropdown = getByLabelText(/select/i) as HTMLDivElement;
    fireEvent.click(dropdown);
    const options = getByTestId('dropdown-options');
    expect(options).toBeInTheDocument();
  });

  test('should set isExpanded to false if click outside of dropdown', () => {
    const { getByLabelText, getByTestId } = render(
      <DropdownMultiselection {...defaultProps} />
    );
    const dropdown = getByLabelText(/select/i) as HTMLDivElement;
    fireEvent.click(dropdown);
    const options = getByTestId('dropdown-options');
    expect(options).toBeInTheDocument();
    fireEvent.mouseUp(document);
    expect(options).not.toBeInTheDocument();
  });

  describe('Keyboard events handling', () => {
    let dropdown: HTMLDivElement;
    let chevron: HTMLImageElement;
    beforeEach(() => {
      const { getByText, getByAltText } = render(
        <DropdownMultiselection {...defaultProps} />
      );
      dropdown = getByText(/select/i) as HTMLDivElement;
      chevron = getByAltText('chevron-icon') as HTMLImageElement;
    });

    test('should set isExpanded = true on `space`', () => {
      fireEvent.keyDown(dropdown, { key: ' ' });
      const style = window.getComputedStyle(chevron);
      expect(style.transform).toBe('rotate(180deg)');
    });

    test('should set isExpanded = true on `enter`', () => {
      fireEvent.keyDown(dropdown, { key: 'Enter' });
      const style = window.getComputedStyle(chevron);
      expect(style.transform).toBe('rotate(180deg)');
    });

    test('should set isExpanded = true on `escape`', () => {
      fireEvent.keyDown(dropdown, { key: 'Escape' });
      const style = window.getComputedStyle(chevron);
      expect(style.transform).toBe('');
    });
  });

  describe('Disabled Dropdown', () => {
    test('should be disabled and not expand on click', () => {
      const { getByLabelText, getByAltText } = render(
        <DropdownMultiselection {...defaultProps} disabled />
      );

      let dropdown = getByLabelText(/select/i) as HTMLDivElement;
      let chevron = getByAltText('chevron-icon') as HTMLImageElement;
      fireEvent.click(dropdown);
      expect(dropdown).toHaveAttribute('disabled');
      const style = window.getComputedStyle(chevron);
      expect(style.transform).toBe('');
    });

    test('should not handle click events', () => {
      const onOptionClickMock = jest.fn();
      const {
        container: { firstChild },
      } = render(
        <DropdownMultiselection
          {...defaultProps}
          disabled={true}
          onOptionClick={onOptionClickMock}
        />
      );

      firstChild && fireEvent.click(firstChild);
      expect(onOptionClickMock).not.toBeCalled();
    });
  });
});
