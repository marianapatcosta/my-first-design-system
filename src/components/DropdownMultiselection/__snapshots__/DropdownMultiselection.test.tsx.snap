// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMultiselection renders without errors and matches snapshot 1`] = `
<div
  class="sc-eCssSg CECUs"
  id=""
>
  <div
    aria-expanded="false"
    aria-label="Select select an option"
    class="sc-jSgupP dnWQpk"
    role="button"
    tabindex="0"
  >
    <div>
      select an option
    </div>
    <img
      alt="chevron-icon"
      class="sc-gKsewC eSggAZ"
      src="chevron.svg"
    />
  </div>
</div>
`;
